torch/autograd/gradcheck.py

```diff
@@ -161,13 +161,54 @@ def get_stride(size):
     return jacobian
 
 
-def check_analytical_jacobian_attributes(inputs, output, nondet_tol, grad_out, check_grad_dtypes,
-                                         raise_exception):
-    jacobian, reentrant, correct_grad_sizes, correct_grad_types = \
-        get_analytical_jacobian(inputs, output, nondet_tol, grad_out)
+def check_jacobians_equal(j1, j2, atol):
+    # Check whether the max diff betwen two jacobians are within some tolerance `atol`
+    for j1_x, j2_x in zip(j1, j2):
+        if j1_x.numel() != 0 and (j1_x - j2_x).abs().max() > atol:
+            return False
+    return True
+
+
+def combine_jacobian_rows(jacobians_rows, inputs, output):
+    out_jacobians = make_jacobian(inputs, output.numel())
+    diff_input_list = list(iter_tensors(inputs, True))
+    correct_grad_sizes = True
+    correct_grad_types = True
+    for i, rows in jacobians_rows.items():
+        inp = diff_input_list[i]
+        out_jacobian = out_jacobians[i]
+        for j, row in enumerate(rows):
+            if row is not None and row.size() != inp.size():
+                correct_grad_sizes = False
+            elif row is not None and row.dtype != inp.dtype:
+                correct_grad_types = False
+            if row is None:
+                out_jacobian[:, j].zero_()
+            else:
+                row_dense = row.to_dense() if not row.layout == torch.strided else row
+                assert out_jacobian[:, j].numel() == row_dense.numel()
+                out_jacobian[:, j] = row_dense.contiguous().view(-1)
+    return out_jacobians, correct_grad_sizes, correct_grad_types
+
+
+def check_analytical_jacobian_attributes(inputs, output, nondet_tol, grad_out_scale, check_grad_dtypes,
+                                         raise_exception, custom_backward_fn=None):
+    diff_input_list = list(iter_tensors(inputs, True))
+
+    def backward_fn(grad_output):
+        return torch.autograd.grad(output, diff_input_list, grad_output,
+                                   retain_graph=True, allow_unused=True)
+    fn = custom_backward_fn if custom_backward_fn is not None else backward_fn
+    jacobians_rows = get_analytical_jacobian(fn, output.clone(), grad_out_scale)
+    jacobians_rows_reentrant = get_analytical_jacobian(fn, output.clone(), grad_out_scale)
```

Comment:
I never really understood what the reentrant check is testing. What is it testing?

The reason why I am asking is because previously, we would compute row 0 of the first jacobian, then row 0 of the second jacobian, then row 1 of the first jacobian, then row 1 of the second jacobian, etc. This PR changes it so that we compute the full first jacobian followed by the full second jacobian. Does this affect the reentrant check?

Reply:
I don't really know what its named 'reentrant' check, AFAICT its just computing something twice and checking if they are the same as to see if its deterministic. A better name imo should be determinism check not reentrant check. Maybe @albanD knows more.

Reply:
Offline @albanD said this is re-entrant in the sense that it is calling the same function twice. This is different from when we call autograd from within autograd, which we also refer to as "re-entrant". So yes, looks like a case of bad naming

Reply:
Also I realized that the code does explain this in an error message: https://github.com/pytorch/pytorch/blob/87989a6cf9979d3842991ab1696a62a1d247a523/torch/autograd/gradcheck.py#L472-L475